### DISCUSSION TITLE
YourTTS with LJSpeech

### DISCUSSION THREAD

**Discussion by icklerly1** [coqui-ai/TTS#2217] [General]
Hi,

I am using your script to train a custom YourTTS model.
-> [train_yourtts.py](https://github.com/coqui-ai/TTS/blob/061ac43187fbcc98ee73e158f672b7fea1e8da47/recipes/vctk/yourtts/train_yourtts.py)

I am just exchanging the config to:
BaseDatasetConfig(
    formatter="ljspeech", dataset_name="rebotnix", meta_file_train=os.path.join(train_dir,"train.csv"), meta_file_val="", path="", language="de"
)

The train.csv has the correct format.

After the training I am trying to do an inference but get the following error.
No speaker_idxs are found.

> Available speaker ids: (Set --speaker_idx flag to one of these values to use the multi-speaker model.
{}

What am I doing wrong?

**Comment by p0p4k** [comment]
During inference, I think you are supposed to provide a speaker index.

**Reply by Ca-ressemble-a-du-fake** [reply]
This is related to this other [question](https://github.com/coqui-ai/TTS/discussions/2330). @p0p4k could you elaborate on "providing a speaker index" ? Looking at [source code] (https://github.com/coqui-ai/TTS/blob/914280a556b06c957f48f18fb64b61b8b984b37d/TTS/bin/synthesize.py#L206) a boolean is expected not a list. providing `--list_speaker_idxs True` does not help, result is still {}.

**Comment by p0p4k** [comment]
Hello @Ca-ressemble-a-du-fake ,
can you give me output for following in your CLI ?
tts --list_speaker_idx --model_name tts_models/en/vctk/vits
Thanks.

**Reply by Ca-ressemble-a-du-fake** [reply]
Here is the output :
``` 
> Available speaker ids: (Set --speaker_idx flag to one of these values to use the multi-speaker model.
{'ED\n': 0, 'p225': 1, 'p226': 2, 'p227': 3, 'p228': 4, 'p229': 5, 'p230': 6, 'p231': 7, 'p232': 8, 'p233': 9, 'p234': 10, 'p236': 11, 'p237': 12, 'p238': 13, 'p239': 14, 'p240': 15, 'p241': 16, 'p243': 17, 'p244': 18, 'p245': 19, 'p246': 20, 'p247': 21, 'p248': 22, 'p249': 23, 'p250': 24, 'p251': 25, 'p252': 26, 'p253': 27, 'p254': 28, 'p255': 29, 'p256': 30, 'p257': 31, 'p258': 32, 'p259': 33, 'p260': 34, 'p261': 35, 'p262': 36, 'p263': 37, 'p264': 38, 'p265': 39, 'p266': 40, 'p267': 41, 'p268': 42, 'p269': 43, 'p270': 44, 'p271': 45, 'p272': 46, 'p273': 47, 'p274': 48, 'p275': 49, 'p276': 50, 'p277': 51, 'p278': 52, 'p279': 53, 'p280': 54, 'p281': 55, 'p282': 56, 'p283': 57, 'p284': 58, 'p285': 59, 'p286': 60, 'p287': 61, 'p288': 62, 'p292': 63, 'p293': 64, 'p294': 65, 'p295': 66, 'p297': 67, 'p298': 68, 'p299': 69, 'p300': 70, 'p301': 71, 'p302': 72, 'p303': 73, 'p304': 74, 'p305': 75, 'p306': 76, 'p307': 77, 'p308': 78, 'p310': 79, 'p311': 80, 'p312': 81, 'p313': 82, 'p314': 83, 'p316': 84, 'p317': 85, 'p318': 86, 'p323': 87, 'p326': 88, 'p329': 89, 'p330': 90, 'p333': 91, 'p334': 92, 'p335': 93, 'p336': 94, 'p339': 95, 'p340': 96, 'p341': 97, 'p343': 98, 'p345': 99, 'p347': 100, 'p351': 101, 'p360': 102, 'p361': 103, 'p362': 104, 'p363': 105, 'p364': 106, 'p374': 107, 'p376': 108}
```
So it does not work only for YourTTS.

If I do the same with with YourTTS :

```
tts --list_speaker_idx --model_path /home/caraduf/Models/YourTTS_ME_16k_With_Names_In_Formatter-February-09-2023_07+02AM-0000000/best_model.pth
```
It yields :

```
Traceback (most recent call last):
  File "/home/caraduf/CoquiTTS/coquienv/bin/tts", line 8, in <module>
    sys.exit(main())
  File "/home/caraduf/CoquiTTS/TTS/TTS/bin/synthesize.py", line 316, in main
    synthesizer = Synthesizer(
  File "/home/caraduf/CoquiTTS/TTS/TTS/utils/synthesizer.py", line 75, in __init__
    self._load_tts(tts_checkpoint, tts_config_path, use_cuda)
  File "/home/caraduf/CoquiTTS/TTS/TTS/utils/synthesizer.py", line 108, in _load_tts
    self.tts_config = load_config(tts_config_path)
  File "/home/caraduf/CoquiTTS/TTS/TTS/config/__init__.py", line 79, in load_config
    ext = os.path.splitext(config_path)[1]
  File "/usr/lib/python3.10/posixpath.py", line 118, in splitext
    p = os.fspath(p)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

So I need to add `--config_path` but then after a very loooooong time (way longer than with your example)  I get `{}` (emplty list)

Edit : it took so much time because the speaker encoder files were pointing to `https://...` for ease of use. Now that I hardcoded to a local path I don't have to wait several minutes! I can do more tests!

**Reply by Ca-ressemble-a-du-fake** [reply]
If I run the above command for the pretrained model it runs picobello : `tts --list_speaker_idx --model_name tts_models/multilingual/multi-dataset/your_tts` 

```
 > Available speaker ids: (Set --speaker_idx flag to one of these values to use the multi-speaker model.
{'female-en-5': 0, 'female-en-5\n': 1, 'female-pt-4\n': 2, 'male-en-2': 3, 'male-en-2\n': 4, 'male-pt-3\n': 5}
```

And in both cases (vctk vits and pretrained yourtts) changing the speakers_id.json / speakers.json filenames produces an error (which is expected). So I tend to think that somewhere in the recipe the speaker dvector are not seen for an unknown reason.

**Comment by Ca-ressemble-a-du-fake** [comment]
Maybe the issue lies in the `speakers_file` format. For vctk VITS the speaker_file points to a json file, in pretrained multilingual YourTTS it's a json file too with the embeddings of the 38 speakers. But what is being generated by this recipe is a binary pth model. 

Yet the speakers_file specified in the config.json does not seem to be used because I made a typo, so file did not exist and there was no error (I would have expected a file not found error). 

I finally tried by providing the speakers_file_path as CLI argument but with the same result (empty list).

Yet it can be noticed that in the config.json generated by the recipe `use_speaker_embedding` was false whereas in the pretrained ~~yourtts model and in~~ vctk vits model it is set to True (which would make sense). But even setting it to true does not make any difference. So it seems that speakers_file / speaker embeddings are not loaded. Maybe @Edresson can confirm my hypothesis ?

Edit : I twisted things because yourtts pretrained model uses null as speakers_file and json file as d-vectors. 

Now if it appears that for the YourTTS recipe `d_vector_file` is None in https://github.com/coqui-ai/TTS/blob/914280a556b06c957f48f18fb64b61b8b984b37d/TTS/config/__init__.py#L129 although config.json has a value for it and there is no typo in it.

**Reply by Ca-ressemble-a-du-fake** [reply]
If I print `config_dict` it still has `d_vector_file` set to the correct value in https://github.com/coqui-ai/TTS/blob/914280a556b06c957f48f18fb64b61b8b984b37d/TTS/config/__init__.py#L92
So it is really there but then it disappears (ie it becomes None) in `config` https://github.com/coqui-ai/TTS/blob/914280a556b06c957f48f18fb64b61b8b984b37d/TTS/config/__init__.py#L96

So there must be something weird happening in between!

**Reply by Ca-ressemble-a-du-fake** [reply]
From my tests it always returns an empty list when a list is used for `d_vector_file`. I tried to use the `speaker.pth` (not between brackets) and I get 
```
Traceback (most recent call last):
  File "/home/caraduf/CoquiTTS/coquienv/bin/tts", line 8, in <module>
    sys.exit(main())
  File "/home/caraduf/CoquiTTS/TTS/TTS/bin/synthesize.py", line 316, in main
    synthesizer = Synthesizer(
  File "/home/caraduf/CoquiTTS/TTS/TTS/utils/synthesizer.py", line 75, in __init__
    self._load_tts(tts_checkpoint, tts_config_path, use_cuda)
  File "/home/caraduf/CoquiTTS/TTS/TTS/utils/synthesizer.py", line 112, in _load_tts
    self.tts_model = setup_tts_model(config=self.tts_config)
  File "/home/caraduf/CoquiTTS/TTS/TTS/tts/models/__init__.py", line 13, in setup_model
    model = MyModel.init_from_config(config, samples)
  File "/home/caraduf/CoquiTTS/TTS/TTS/tts/models/vits.py", line 1743, in init_from_config
    speaker_manager = SpeakerManager.init_from_config(config, samples)
  File "/home/caraduf/CoquiTTS/TTS/TTS/tts/utils/speakers.py", line 117, in init_from_config
    speaker_manager = SpeakerManager(
  File "/home/caraduf/CoquiTTS/TTS/TTS/tts/utils/speakers.py", line 63, in __init__
    super().__init__(
  File "/home/caraduf/CoquiTTS/TTS/TTS/tts/utils/managers.py", line 151, in __init__
    self.load_embeddings_from_file(embedding_file_path)
  File "/home/caraduf/CoquiTTS/TTS/TTS/tts/utils/managers.py", line 212, in load_embeddings_from_file
    self.name_to_id, self.clip_ids, self.embeddings, self.embeddings_by_names = self.read_embeddings_from_file(
  File "/home/caraduf/CoquiTTS/TTS/TTS/tts/utils/managers.py", line 194, in read_embeddings_from_file
    speakers = sorted({x["name"] for x in embeddings.values()})
  File "/home/caraduf/CoquiTTS/TTS/TTS/tts/utils/managers.py", line 194, in <setcomp>
    speakers = sorted({x["name"] for x in embeddings.values()})
TypeError: 'int' object is not subscriptable

```

~~So it looks like "pth" is not the expected file. I will try to write a json file with the embeddings and try again.~~(this is not true see below)

**Comment by Ca-ressemble-a-du-fake** [comment]
It looks like using json for speakers embeddings solve the issue. So here instead of `speakers.pth` use `speaker.json`. It will output speaker embeddings as json file which can then be parsed. 
https://github.com/coqui-ai/TTS/blob/914280a556b06c957f48f18fb64b61b8b984b37d/recipes/vctk/yourtts/train_yourtts.py#L95

I have to check again because it works during training (config shows the expected list) but not when using list speaker idxs on the generated best model.

**Comment by Ca-ressemble-a-du-fake** [comment]
Now I got it working : the recipe works as is for the training (so d vectors are referenced in a list).
But then for the inference part, it expects a single d vector file that gathers all speakers. So you have to compute the embeddings with the compute_embeddings script like :
https://github.com/coqui-ai/TTS/blob/3b1a28fa95677fc05efdedb18cfe70b3163339f2/TTS/bin/compute_embeddings.py#L97

It will produce a .pth (or a .json if you give --output_path) that you have to reference in the config.json file that you give to the `tts --config_path` command in the sections (caution there are 2 places) : `d_vector_file = "path_to_the_single_big_speakers.pth"`

**Reply by Ca-ressemble-a-du-fake** [reply]
Actually this is already explained in the synthetize.py source code for multipspeaker models https://github.com/coqui-ai/TTS/blob/661725b95e1d50aa104395fdab965ae000665e7c/TTS/bin/synthesize.py#L101

So you don't need to change anything in the config.json you simply can pass  `--speakers_file_path path/to/speaker.json`. The help text could add something along those lines `"speakers_file_path" file must be computed by script bin/compute_embeddings.py`.
